torch/lib/c10d/TCPStore.hpp

```diff
@@ -36,29 +70,55 @@ class TCPStoreDaemon {
   void getNumKeysHandler(int socket) const;
   void deleteHandler(int socket);
   void waitHandler(int socket);
+  void watchHandler(int socket);
 
   bool checkKeys(const std::vector<std::string>& keys) const;
+  // Helper function to alerts waiting workers, used in setHandler, getHandler
   void wakeupWaitingClients(const std::string& key);
+  // Helper function used when the key is changed
+  // used in setHandler, addHandler, getHandler, deleteHandler
+  void sendKeyUpdatesToClients(
+      const std::string& key,
+      const enum WatchResponseType& type,
+      std::vector<uint8_t>& oldData,
+      std::vector<uint8_t>& newData);
 
-  void initStopSignal();
-  void closeStopSignal();
-
-  std::thread daemonThread_;
+ private:
   std::unordered_map<std::string, std::vector<uint8_t>> tcpStore_;
-  // From key -> the list of sockets waiting on it
+  // From key -> the list of sockets waiting on the key
   std::unordered_map<std::string, std::vector<int>> waitingSockets_;
   // From socket -> number of keys awaited
   std::unordered_map<int, size_t> keysAwaited_;
+  // From key -> the list of sockets watching the key
+  std::unordered_map<std::string, std::vector<int>> watchedSockets_;
+};
 
-  std::vector<int> sockets_;
-  int storeListenSocket_;
-#ifdef _WIN32
-  const std::chrono::milliseconds checkTimeout_
-      = std::chrono::milliseconds(10);
-  HANDLE ghStopEvent_;
-#else
-  std::vector<int> controlPipeFd_{-1, -1};
-#endif
+// Separate thread that is launched on all instances (including master)
+// Right now only handles callbacks registered from watchKey()
+class TCPStoreWorkerDaemon : public BackgroundThread {
+ public:
+  explicit TCPStoreWorkerDaemon(int listenSocket);
+  // Adds a callback to run key change
+  void addCallback(
+      std::string key,
+      std::function<
+          void(c10::optional<std::string>, c10::optional<std::string>)> cb);
+  std::mutex mtx;
+  std::condition_variable cv;
+  bool callbackRegistered = false;
```

Comment:
it's a bit weird to have public fields like this. IIUC, this is to allow `TCPStore` to easily manipulate `TCPStoreWorkerDaemon` thread states. Is it possible to wrap those into `TCPStoreWorkerDaemon` member functions so that we don't need to expose its raw internal states.